jbang --java 17 when default is java 11 build integration fails

**Describe the bug**
Have java 11 in system path.
Run `jbang --java 17 https://gist.github.com/rivasdiaz/f2e9efd1bf2079f0101f865846af2502` and you get error about 

```
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ConfigBuildStep#generateConfigClasses threw an exception: java.lang.UnsupportedClassVersionError: MavenStatsConfig has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:445)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:405)
        at io.quarkus.deployment.configuration.ConfigMappingUtils.toClass(ConfigMappingUtils.java:93)
        at io.quarkus.deployment.configuration.ConfigMappingUtils.generateConfigClasses(ConfigMappingUtils.java:51)
```

as if it compiles with java 17 but then falls back to running quarkus build integration with java 11...

solution would be to run that integration externally when jbang is running with lower version.

or at least help with info to user on what is happening.

--- a/src/main/java/dev/jbang/source/Source.java
+++ b/src/main/java/dev/jbang/source/Source.java
@@ -88,11 +88,17 @@ public abstract class Source {
 		return contents;
 	}
 
-	public List<String> getLines() {
+	public Stream<String> getLines() {
 		if (lines == null && contents != null) {
 			lines = Arrays.asList(contents.split("\\r?\\n"));
 		}
-		return lines;
+		return lines.stream();
+	}
+
+	public Stream<String> getTags() {
+		// TODO: Insert `.takeWhile(s -> Util.isBlankString(s) || s.startsWith("//"))`
+		// once we support at least Java 9
+		return getLines().filter(s -> s.startsWith("//"));
 	}
 
 	public Optional<String> getJavaPackage() {
@@ -104,13 +110,7 @@ public abstract class Source {
 	}
 
 	public List<String> collectDependencies() {
-		// Make sure that dependencies declarations are well formatted
-		if (getLines().stream().anyMatch(it -> it.startsWith("// DEPS"))) {
-			throw new IllegalArgumentException("Dependencies must be declared by using the line prefix //DEPS");
-		}
-
-		return getLines()	.stream()
-							.filter(Source::isDependDeclare)
+		return getLines()	.filter(Source::isDependDeclare)
 							.flatMap(Source::extractDependencies)
 							.map(replaceProperties)
 							.collect(Collectors.toList());
@@ -196,8 +196,7 @@ public abstract class Source {
 	}
 
 	public List<MavenRepo> collectRepositories() {
-		return getLines()	.stream()
-							.filter(Source::isRepoDeclare)
+		return getLines()	.filter(Source::isRepoDeclare)
 							.flatMap(Source::extractRepositories)
 							.map(replaceProperties)
 							.map(DependencyUtil::toMavenRepo)
@@ -240,8 +239,7 @@ public abstract class Source {
 	}
 
 	public Optional<String> getDescription() {
-		String desc = getLines().stream()
-								.filter(Source::isDescriptionDeclare)
+		String desc = getTags()	.filter(Source::isDescriptionDeclare)
 								.map(s -> s.substring(DESCRIPTION_COMMENT_PREFIX.length()))
 								.collect(Collectors.joining("\n"));
 		if (desc.isEmpty()) {
@@ -256,8 +254,7 @@ public abstract class Source {
 	}
 
 	public Optional<String> getGav() {
-		List<String> gavs = getLines()	.stream()
-										.filter(Source::isGavDeclare)
+		List<String> gavs = getTags()	.filter(Source::isGavDeclare)
 										.map(s -> s.substring(GAV_COMMENT_PREFIX.length()))
 										.collect(Collectors.toList());
 		if (gavs.isEmpty()) {
@@ -290,15 +287,12 @@ public abstract class Source {
 
 	private List<String> collectRawOptions(String prefix) {
 		String joptsPrefix = "//" + prefix;
-
-		List<String> lines = getLines();
-
-		List<String> javaOptions = lines.stream()
-										.map(it -> it.split(" // ")[0]) // strip away nested comments.
-										.filter(it -> it.startsWith(joptsPrefix + " ")
-												|| it.startsWith(joptsPrefix + "\t") || it.equals(joptsPrefix))
-										.map(it -> it.replaceFirst(joptsPrefix, "").trim())
-										.collect(Collectors.toList());
+		List<String> javaOptions = getTags()
+											.map(it -> it.split(" // ")[0]) // strip away nested comments.
+											.filter(it -> it.startsWith(joptsPrefix + " ")
+													|| it.startsWith(joptsPrefix + "\t") || it.equals(joptsPrefix))
+											.map(it -> it.replaceFirst(joptsPrefix, "").trim())
+											.collect(Collectors.toList());
 
 		String envOptions = System.getenv("JBANG_" + prefix);
 		if (envOptions != null) {
@@ -327,14 +321,13 @@ public abstract class Source {
 	}
 
 	public List<RefTarget> collectFiles(ResourceResolver siblingResolver) {
-		return getLines()	.stream()
-							.filter(f -> f.startsWith(FILES_COMMENT_PREFIX))
-							.flatMap(line -> Arrays	.stream(line.split(" // ")[0].split("[ ;,]+"))
-													.skip(1)
-													.map(String::trim))
-							.map(replaceProperties)
-							.map(f -> toFileRef(f, siblingResolver))
-							.collect(Collectors.toCollection(ArrayList::new));
+		return getTags().filter(f -> f.startsWith(FILES_COMMENT_PREFIX))
+						.flatMap(line -> Arrays	.stream(line.split(" // ")[0].split("[ ;,]+"))
+												.skip(1)
+												.map(String::trim))
+						.map(replaceProperties)
+						.map(f -> toFileRef(f, siblingResolver))
+						.collect(Collectors.toCollection(ArrayList::new));
 	}
 
 	private RefTarget toFileRef(String fileReference, ResourceResolver siblingResolver) {
@@ -346,21 +339,20 @@ public abstract class Source {
 	}
 
 	public List<Source> collectSources(ResourceResolver siblingResolver) {
-		if (getLines() == null) {
+		if (getContents() == null) {
 			return Collections.emptyList();
 		} else {
 			String org = getResourceRef().getOriginalResource();
 			Path baseDir = org != null ? getResourceRef().getFile().getAbsoluteFile().getParentFile().toPath()
 					: Util.getCwd();
-			return getLines()	.stream()
-								.filter(f -> f.startsWith(SOURCES_COMMENT_PREFIX))
-								.flatMap(line -> Arrays	.stream(line.split(" // ")[0].split("[ ;,]+"))
-														.skip(1)
-														.map(String::trim))
-								.map(replaceProperties)
-								.flatMap(line -> Util.explode(org, baseDir, line).stream())
-								.map(ref -> forResource(siblingResolver, ref, replaceProperties))
-								.collect(Collectors.toCollection(ArrayList::new));
+			return getTags().filter(f -> f.startsWith(SOURCES_COMMENT_PREFIX))
+							.flatMap(line -> Arrays	.stream(line.split(" // ")[0].split("[ ;,]+"))
+													.skip(1)
+													.map(String::trim))
+							.map(replaceProperties)
+							.flatMap(line -> Util.explode(org, baseDir, line).stream())
+							.map(ref -> forResource(siblingResolver, ref, replaceProperties))
+							.collect(Collectors.toCollection(ArrayList::new));
 		}
 	}
 

--- a/src/main/java/dev/jbang/source/builders/BaseBuilder.java
+++ b/src/main/java/dev/jbang/source/builders/BaseBuilder.java
@@ -162,7 +162,7 @@ public abstract class BaseBuilder implements Builder {
 			System.setProperty(entry.getKey(), entry.getValue());
 		}
 		IntegrationResult integrationResult = IntegrationManager.runIntegrations(ss,
-				compileDir.toPath(), pomPath, ctx.isNativeImage());
+				compileDir.toPath(), pomPath, ctx.isNativeImage(), requestedJavaVersion);
 		System.setProperties(old);
 
 		if (ctx.getMainClass() == null) { // if non-null user forced set main

--- a/src/main/java/dev/jbang/source/builders/BaseBuilder.java
+++ b/src/main/java/dev/jbang/source/builders/BaseBuilder.java
@@ -161,10 +161,8 @@ public abstract class BaseBuilder implements Builder {
 		for (Map.Entry<String, String> entry : ctx.getProperties().entrySet()) {
 			System.setProperty(entry.getKey(), entry.getValue());
 		}
-		IntegrationResult integrationResult = IntegrationManager.runIntegration(ss.getRepositories(),
-				ss.getClassPath().getArtifacts(),
-				compileDir.toPath(), pomPath,
-				ss.getMainSource(), ctx.isNativeImage());
+		IntegrationResult integrationResult = IntegrationManager.runIntegrations(ss,
+				compileDir.toPath(), pomPath, ctx.isNativeImage());
 		System.setProperties(old);
 
 		if (ctx.getMainClass() == null) { // if non-null user forced set main

--- a/src/main/java/dev/jbang/spi/IntegrationInput.java
+++ b/src/main/java/dev/jbang/spi/IntegrationInput.java
@@ -1,0 +1,33 @@
+package dev.jbang.spi;
+
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+
+import com.google.gson.annotations.SerializedName;
+
+public class IntegrationInput {
+	@SerializedName(value = "integration")
+	public final String integrationClassName;
+	public final Path source;
+	public final Path classes;
+	public final Path pom;
+	public final Map<String, String> repositories;
+	public final Map<String, Path> dependencies;
+	public final List<String> comments;
+	@SerializedName(value = "native")
+	public final boolean nativeRequested;
+
+	public IntegrationInput(String integrationClassName, Path source, Path classes, Path pom,
+			Map<String, String> repositories, Map<String, Path> dependencies, List<String> comments,
+			boolean nativeRequested) {
+		this.integrationClassName = integrationClassName;
+		this.source = source;
+		this.classes = classes;
+		this.pom = pom;
+		this.repositories = repositories;
+		this.dependencies = dependencies;
+		this.comments = comments;
+		this.nativeRequested = nativeRequested;
+	}
+}

--- a/src/main/java/dev/jbang/spi/IntegrationManager.java
+++ b/src/main/java/dev/jbang/spi/IntegrationManager.java
@@ -20,10 +20,13 @@ import java.nio.file.Path;
 import java.util.*;
 import java.util.stream.Collectors;
 
+import javax.annotation.Nonnull;
+
 import dev.jbang.cli.ExitException;
 import dev.jbang.dependencies.ArtifactInfo;
 import dev.jbang.dependencies.MavenRepo;
 import dev.jbang.source.Source;
+import dev.jbang.source.SourceSet;
 import dev.jbang.util.Util;
 
 /**
@@ -41,101 +44,37 @@ public class IntegrationManager {
 	 * <p>
 	 * If an integration point created a native image it returns the resulting
 	 * image.
-	 *
-	 *
-	 * @param repositories
-	 * @param artifacts
-	 * @param tmpJarDir
-	 * @param pomPath
-	 * @param source
-	 * @param nativeRequested
-	 * @return
 	 */
-	public static IntegrationResult runIntegration(List<MavenRepo> repositories, List<ArtifactInfo> artifacts,
-			Path tmpJarDir,
-			Path pomPath, Source source,
+	public static IntegrationResult runIntegrations(SourceSet ss, Path tmpJarDir, Path pomPath,
 			boolean nativeRequested) {
-		URL[] urls = artifacts.stream().map(s -> {
-			try {
-				return s.getFile().toURI().toURL();
-			} catch (MalformedURLException e) {
-				throw new RuntimeException(e);
-			}
-		}).toArray(URL[]::new);
-		List<String> comments = source.getLines().stream().filter(s -> s.startsWith("//")).collect(Collectors.toList());
-		URLClassLoader integrationCl = new URLClassLoader(urls);
+		IntegrationResult result = new IntegrationResult(null, null, null);
+		Source source = ss.getMainSource();
+
+		LinkedHashMap<String, String> repos = new LinkedHashMap<>();
+		LinkedHashMap<String, Path> deps = new LinkedHashMap<>();
+		for (MavenRepo repo : ss.getRepositories()) {
+			repos.put(repo.getId(), repo.getUrl());
+		}
+		for (ArtifactInfo art : ss.getClassPath().getArtifacts()) {
+			deps.put(art.getCoordinate().toCanonicalForm(), art.getFile().toPath());
+		}
+
+		List<String> comments = source.getTags().collect(Collectors.toList());
 		ClassLoader old = Thread.currentThread().getContextClassLoader();
-		Map<String, byte[]> data = new HashMap<>();
-		List<Map.Entry<String, String>> repos = null;
-		List<Map.Entry<String, Path>> deps = null;
-		Path nativeImage = null;
-		String mainClass = null;
-		List<String> javaArgs = null;
 		PrintStream oldout = System.out;
 		try {
-			// TODO: should we add new properties to the integration method?
-			if (source.getResourceRef().getFile() != null) {
-				System.setProperty("jbang.source", source.getResourceRef().getFile().getAbsolutePath());
-			}
+			URLClassLoader integrationCl = getClassLoader(deps);
 			Thread.currentThread().setContextClassLoader(integrationCl);
 			Set<String> classNames = loadIntegrationClassNames(integrationCl);
 			for (String className : classNames) {
-				if (repos == null) {
-					repos = repositories.stream()
-										.map(s -> new AbstractMap.SimpleEntry<>(s.getId(), s.getUrl()))
-										.collect(Collectors.toList());
-				}
-				if (deps == null) {
-					deps = artifacts.stream()
-									.map(s -> new AbstractMap.SimpleEntry<>(s.getCoordinate().toCanonicalForm(),
-											s.getFile().toPath()))
-									.collect(Collectors.toList());
-				}
-				Class<?> clazz = Class.forName(className, true, integrationCl);
-				Method method = clazz.getDeclaredMethod("postBuild", Path.class, Path.class, List.class, List.class,
-						List.class,
-						boolean.class);
-				Util.infoMsg("Post build with " + className);
-
-				if (Util.isVerbose()) {
-					System.setOut(new PrintStream(new FileOutputStream(FileDescriptor.err)));
-				} else {
-					System.setOut(new PrintStream(new OutputStream() {
-						public void write(int b) {
-							// DO NOTHING
-							// TODO: capture it for later print if error
-						}
-					}));
-				}
-
-				@SuppressWarnings("unchecked")
-				Map<String, Object> integrationResult = (Map<String, Object>) method.invoke(null, tmpJarDir, pomPath,
-						repos, deps, comments, nativeRequested);
-				@SuppressWarnings("unchecked")
-				Map<String, byte[]> ret = (Map<String, byte[]>) integrationResult.get(FILES);
-				if (ret != null) {
-					data.putAll(ret);
-				}
-				Path image = (Path) integrationResult.get(NATIVE_IMAGE);
-				if (image != null) {
-					nativeImage = image;
-				}
-				String mc = (String) integrationResult.get(MAIN_CLASS);
-				if (mc != null) {
-					mainClass = mc;
-				}
-				@SuppressWarnings("unchecked")
-				List<String> ja = (List<String>) integrationResult.get(JAVA_ARGS);
-				if (ja != null) {
-					javaArgs = ja;
-				}
-			}
-			for (Map.Entry<String, byte[]> entry : data.entrySet()) {
-				Path target = tmpJarDir.resolve(entry.getKey());
-				Files.createDirectories(target.getParent());
-				try (OutputStream out = Files.newOutputStream(target)) {
-					out.write(entry.getValue());
-				}
+				Path srcPath = (source.getResourceRef().getFile() != null)
+						? source.getResourceRef().getFile().toPath().toAbsolutePath()
+						: null;
+				IntegrationInput input = new IntegrationInput(className, srcPath, tmpJarDir, pomPath, repos, deps,
+						comments,
+						nativeRequested);
+				IntegrationResult ir = runIntegration(integrationCl, input);
+				result = result.merged(ir);
 			}
 		} catch (ClassNotFoundException e) {
 			throw new ExitException(EXIT_UNEXPECTED_STATE, "Unable to load integration class", e);
@@ -150,8 +89,19 @@ public class IntegrationManager {
 			Thread.currentThread().setContextClassLoader(old);
 			System.setOut(oldout);
 		}
-		return new IntegrationResult(nativeImage, mainClass, javaArgs);
+		return result;
+	}
 
+	@Nonnull
+	private static URLClassLoader getClassLoader(Map<String, Path> deps) {
+		URL[] urls = deps.values().stream().map(path -> {
+			try {
+				return path.toUri().toURL();
+			} catch (MalformedURLException e) {
+				throw new RuntimeException(e);
+			}
+		}).toArray(URL[]::new);
+		return new URLClassLoader(urls);
 	}
 
 	private static Set<String> loadIntegrationClassNames(URLClassLoader integrationCl) throws IOException {
@@ -180,4 +130,67 @@ public class IntegrationManager {
 		return classNames;
 	}
 
+	private static IntegrationResult runIntegration(URLClassLoader integrationCl, IntegrationInput input)
+			throws Exception {
+		Util.infoMsg("Post build with " + input.integrationClassName);
+
+		if (input.source != null) {
+			// TODO: should we add new properties to the integration method?
+			System.setProperty("jbang.source", input.source.toString());
+		}
+
+		Class<?> clazz = Class.forName(input.integrationClassName, true, integrationCl);
+		Method method = clazz.getDeclaredMethod("postBuild", Path.class, Path.class, List.class, List.class,
+				List.class, boolean.class);
+
+		if (Util.isVerbose()) {
+			System.setOut(new PrintStream(new FileOutputStream(FileDescriptor.err)));
+		} else {
+			System.setOut(new PrintStream(new OutputStream() {
+				public void write(int b) {
+					// DO NOTHING
+					// TODO: capture it for later print if error
+				}
+			}));
+		}
+
+		@SuppressWarnings("unchecked")
+		Map<String, Object> result = (Map<String, Object>) method.invoke(null, input.classes, input.pom,
+				mapToList(input.repositories), mapToList(input.dependencies), input.comments, input.nativeRequested);
+
+		@SuppressWarnings("unchecked")
+		Map<String, byte[]> files = (Map<String, byte[]>) result.get(FILES);
+		if (files != null) {
+			for (Map.Entry<String, byte[]> entry : files.entrySet()) {
+				Path target = input.classes.resolve(entry.getKey());
+				Files.createDirectories(target.getParent());
+				try (OutputStream out = Files.newOutputStream(target)) {
+					out.write(entry.getValue());
+				}
+			}
+		}
+
+		Path nativeImage = null;
+		String mainClass = null;
+		List<String> javaArgs = null;
+		Path image = (Path) result.get(NATIVE_IMAGE);
+		if (image != null) {
+			nativeImage = image;
+		}
+		String mc = (String) result.get(MAIN_CLASS);
+		if (mc != null) {
+			mainClass = mc;
+		}
+		@SuppressWarnings("unchecked")
+		List<String> ja = (List<String>) result.get(JAVA_ARGS);
+		if (ja != null) {
+			javaArgs = ja;
+		}
+
+		return new IntegrationResult(nativeImage, mainClass, javaArgs);
+	}
+
+	private static <K, V> List<Map.Entry<K, V>> mapToList(Map<K, V> map) {
+		return new ArrayList<>(map.entrySet());
+	}
 }

--- a/src/main/java/dev/jbang/spi/IntegrationManager.java
+++ b/src/main/java/dev/jbang/spi/IntegrationManager.java
@@ -1,15 +1,9 @@
 package dev.jbang.spi;
 
-import static dev.jbang.cli.BaseCommand.EXIT_UNEXPECTED_STATE;
+import static dev.jbang.cli.BaseCommand.*;
+import static dev.jbang.util.JavaUtil.resolveInJavaHome;
 
-import java.io.BufferedReader;
-import java.io.FileDescriptor;
-import java.io.FileOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.io.OutputStream;
-import java.io.PrintStream;
+import java.io.*;
 import java.lang.reflect.Method;
 import java.net.MalformedURLException;
 import java.net.URL;
@@ -17,10 +11,18 @@ import java.net.URLClassLoader;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.*;
 import java.util.stream.Collectors;
 
 import javax.annotation.Nonnull;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
 
 import dev.jbang.cli.ExitException;
 import dev.jbang.dependencies.ArtifactInfo;
@@ -33,11 +35,12 @@ import dev.jbang.util.Util;
  * JBang uses a 'convention based interface' for build time integration.
  */
 public class IntegrationManager {
-
 	public static final String FILES = "files";
 	public static final String NATIVE_IMAGE = "native-image";
 	public static final String MAIN_CLASS = "main-class";
 	public static final String JAVA_ARGS = "java-args";
+
+	private static final GsonBuilder gsonb = new GsonBuilder().registerTypeAdapter(Path.class, new PathTypeAdapter());
 
 	/**
 	 * Discovers all integration points and runs them.
@@ -46,7 +49,7 @@ public class IntegrationManager {
 	 * image.
 	 */
 	public static IntegrationResult runIntegrations(SourceSet ss, Path tmpJarDir, Path pomPath,
-			boolean nativeRequested) {
+			boolean nativeRequested, String requestedJavaVersion) {
 		IntegrationResult result = new IntegrationResult(null, null, null);
 		Source source = ss.getMainSource();
 
@@ -73,18 +76,20 @@ public class IntegrationManager {
 				IntegrationInput input = new IntegrationInput(className, srcPath, tmpJarDir, pomPath, repos, deps,
 						comments,
 						nativeRequested);
-				IntegrationResult ir = runIntegration(integrationCl, input);
+				IntegrationResult ir = requestedJavaVersion == null
+						? runIntegrationEmbedded(input, integrationCl)
+						: runIntegrationExternal(input, requestedJavaVersion);
 				result = result.merged(ir);
 			}
 		} catch (ClassNotFoundException e) {
-			throw new ExitException(EXIT_UNEXPECTED_STATE, "Unable to load integration class", e);
+			throw new ExitException(EXIT_INVALID_INPUT, "Unable to load integration class", e);
 		} catch (NoSuchMethodException e) {
 			throw new ExitException(
-					EXIT_UNEXPECTED_STATE,
+					EXIT_INVALID_INPUT,
 					"Integration class missing method with signature public static Map<String, byte[]> postBuild(Path classesDir, Path pomFile, List<Map.Entry<String, Path>> dependencies)",
 					e);
 		} catch (Exception e) {
-			throw new ExitException(EXIT_UNEXPECTED_STATE, "Issue running postBuild()", e);
+			throw new ExitException(EXIT_GENERIC_ERROR, "Issue running postBuild()", e);
 		} finally {
 			Thread.currentThread().setContextClassLoader(old);
 			System.setOut(oldout);
@@ -130,7 +135,7 @@ public class IntegrationManager {
 		return classNames;
 	}
 
-	private static IntegrationResult runIntegration(URLClassLoader integrationCl, IntegrationInput input)
+	private static IntegrationResult runIntegrationEmbedded(IntegrationInput input, URLClassLoader integrationCl)
 			throws Exception {
 		Util.infoMsg("Post build with " + input.integrationClassName);
 
@@ -190,7 +195,99 @@ public class IntegrationManager {
 		return new IntegrationResult(nativeImage, mainClass, javaArgs);
 	}
 
+	private static IntegrationResult runIntegrationExternal(IntegrationInput input, String requestedJavaVersion)
+			throws Exception {
+		Gson parser = gsonb.create();
+		Util.infoMsg("Running external post build for " + input.integrationClassName);
+
+		List<String> args = new ArrayList<>();
+		args.add(resolveInJavaHome("java", requestedJavaVersion)); // TODO
+		args.add("-cp");
+		args.add(Util.getJarLocation().toString());
+		args.add("dev.jbang.spi.IntegrationManager");
+
+		if (Util.isVerbose()) {
+			Util.verboseMsg("Running: " + String.join(" ", args));
+			Util.verboseMsg("Input: " + parser.toJson(input));
+		}
+
+		Process process = new ProcessBuilder(args)
+													.redirectError(ProcessBuilder.Redirect.INHERIT)
+													.start();
+
+		try (Writer w = new BufferedWriter(new OutputStreamWriter(process.getOutputStream()))) {
+			parser.toJson(input, w);
+		}
+
+		String cmdOutput;
+		try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+			cmdOutput = br.lines().collect(Collectors.joining());
+		}
+
+		try {
+			process.waitFor();
+		} catch (InterruptedException e) {
+			throw new ExitException(EXIT_GENERIC_ERROR, "External post build was interrupted", e);
+		}
+
+		Util.verboseMsg("Output: #" + process.exitValue() + " - " + cmdOutput);
+
+		if (process.exitValue() == 0) {
+			return parser.fromJson(new StringReader(cmdOutput), IntegrationResult.class);
+		} else {
+			throw new ExitException(EXIT_GENERIC_ERROR, "External post exited with error: " + cmdOutput);
+		}
+	}
+
 	private static <K, V> List<Map.Entry<K, V>> mapToList(Map<K, V> map) {
 		return new ArrayList<>(map.entrySet());
+	}
+
+	private static class PathTypeAdapter extends TypeAdapter<Path> {
+		@Override
+		public Path read(JsonReader in) throws IOException {
+			if (in.peek() == JsonToken.NULL) {
+				in.nextNull();
+				return null;
+			}
+			String path = in.nextString();
+			return Paths.get(path);
+		}
+
+		@Override
+		public void write(JsonWriter out, Path path) throws IOException {
+			if (path == null) {
+				out.nullValue();
+				return;
+			}
+			out.value(path.toString());
+		}
+	}
+
+	public static void main(String... args) {
+		Gson parser = gsonb.create();
+		IntegrationInput input = parser.fromJson(new InputStreamReader(System.in), IntegrationInput.class);
+		ClassLoader old = Thread.currentThread().getContextClassLoader();
+		PrintStream oldout = System.out;
+		String output = "";
+		boolean ok = false;
+		try {
+			URLClassLoader integrationCl = getClassLoader(input.dependencies);
+			Thread.currentThread().setContextClassLoader(integrationCl);
+			IntegrationResult result = runIntegrationEmbedded(input, integrationCl);
+			output = parser.toJson(result);
+			ok = true;
+		} catch (ClassNotFoundException e) {
+			output = "Unable to load integration class";
+		} catch (NoSuchMethodException e) {
+			output = "Integration class missing method with signature public static Map<String, byte[]> postBuild(Path classesDir, Path pomFile, List<Map.Entry<String, Path>> dependencies)";
+		} catch (Exception e) {
+			output = "Issue running postBuild()";
+		} finally {
+			Thread.currentThread().setContextClassLoader(old);
+			System.setOut(oldout);
+		}
+		System.out.println(output);
+		System.exit(ok ? 0 : 1);
 	}
 }

--- a/src/main/java/dev/jbang/spi/IntegrationManager.java
+++ b/src/main/java/dev/jbang/spi/IntegrationManager.java
@@ -17,12 +17,7 @@ import java.net.URLClassLoader;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 import java.util.stream.Collectors;
 
 import dev.jbang.cli.ExitException;
@@ -87,12 +82,12 @@ public class IntegrationManager {
 			for (String className : classNames) {
 				if (repos == null) {
 					repos = repositories.stream()
-										.map(s -> new MapRepoEntry(s.getId(), s.getUrl()))
+										.map(s -> new AbstractMap.SimpleEntry<>(s.getId(), s.getUrl()))
 										.collect(Collectors.toList());
 				}
 				if (deps == null) {
 					deps = artifacts.stream()
-									.map(s -> new MapEntry(s.getCoordinate().toCanonicalForm(),
+									.map(s -> new AbstractMap.SimpleEntry<>(s.getCoordinate().toCanonicalForm(),
 											s.getFile().toPath()))
 									.collect(Collectors.toList());
 				}
@@ -183,77 +178,6 @@ public class IntegrationManager {
 			}
 		}
 		return classNames;
-	}
-
-	private static class MapEntry implements Map.Entry<String, Path> {
-		private final String key;
-		private Path value;
-
-		private MapEntry(String key, Path value) {
-			this.key = key;
-			this.value = value;
-		}
-
-		@Override
-		public String getKey() {
-			return key;
-		}
-
-		@Override
-		public Path getValue() {
-			return value;
-		}
-
-		@Override
-		public Path setValue(Path value) {
-			Path old = this.value;
-			this.value = value;
-			return old;
-		}
-
-		@Override
-		public String toString() {
-			return "MapEntry{" +
-					"key='" + key + '\'' +
-					", value=" + value +
-					'}';
-		}
-	}
-
-	private static class MapRepoEntry implements Map.Entry<String, String> {
-		private final String key;
-		private String value;
-
-		private MapRepoEntry(String key, String value) {
-			this.key = key;
-			this.value = value;
-		}
-
-		@Override
-		public String getKey() {
-			return key;
-		}
-
-		@Override
-		public String getValue() {
-			return value;
-		}
-
-		@Override
-		public String setValue(String value) {
-			String old = this.value;
-			this.value = value;
-			return old;
-		}
-
-		@Override
-		public String toString() {
-			return "MapEntry{" +
-					"key='" + key + '\'' +
-					", value=" + value +
-					'}';
-
-		}
 	}
 
 }

--- a/src/main/java/dev/jbang/spi/IntegrationResult.java
+++ b/src/main/java/dev/jbang/spi/IntegrationResult.java
@@ -14,4 +14,17 @@ public class IntegrationResult {
 		this.mainClass = mainClass;
 		this.javaArgs = javaArgs;
 	}
+
+	public IntegrationResult merged(IntegrationResult ir) {
+		if (ir.nativeImagePath == null && ir.mainClass == null && ir.javaArgs == null) {
+			return this;
+		} else if (nativeImagePath == null && mainClass == null && javaArgs == null) {
+			return ir;
+		} else {
+			return new IntegrationResult(
+					nativeImagePath != null ? nativeImagePath : ir.nativeImagePath,
+					mainClass != null ? mainClass : ir.mainClass,
+					javaArgs != null ? javaArgs : ir.javaArgs);
+		}
+	}
 }

